nifi-extension-bundles/nifi-iceberg-bundle/nifi-iceberg-processors/src/main/java/org/apache/nifi/processors/iceberg/PutIcebergRecord.java

```diff
@@ -0,0 +1,259 @@
+
+    private volatile IcebergWriter icebergWriter;
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return properties;
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    @OnScheduled
+    public void onScheduled(final ProcessContext context) {
+        final IcebergCatalog icebergCatalog = context.getProperty(ICEBERG_CATALOG).asControllerService(IcebergCatalog.class);
+        catalog = icebergCatalog.getCatalog();
+        icebergWriter = context.getProperty(ICEBERG_WRITER).asControllerService(IcebergWriter.class);
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        final TableIdentifierFlowFileFilter flowFileFilter = new TableIdentifierFlowFileFilter(context, MAXIMUM_BYTES);
+        final List<FlowFile> flowFiles = session.get(flowFileFilter);
+        if (flowFiles.isEmpty()) {
+            return;
+        }
+
+        final TableIdentifier tableIdentifier = flowFileFilter.getTableIdentifier();
+        final RecordReaderFactory recordReaderFactory = context.getProperty(RECORD_READER).asControllerService(RecordReaderFactory.class);
+        processFlowFiles(session, flowFiles, tableIdentifier, recordReaderFactory);
+    }
+
+    private void processFlowFiles(final ProcessSession session, final List<FlowFile> flowFiles, final TableIdentifier tableIdentifier, final RecordReaderFactory recordReaderFactory) {
+        final long started = clock.millis();
+
+        final AtomicReference<Relationship> relationship = new AtomicReference<>(SUCCESS);
+
+        final Table table = getTable(tableIdentifier);
+        final Schema schema = table.schema();
+        final Types.StructType struct = schema.asStruct();
+        final IcebergRowWriter rowWriter = icebergWriter.getRowWriter(table);
+
+        for (final FlowFile flowFile : flowFiles) {
+            try (
+                    InputStream inputStream = session.read(flowFile);
+                    RecordReader recordReader = recordReaderFactory.createRecordReader(flowFile, inputStream, getLogger())
+            ) {
+                final AtomicLong recordsProcessed = new AtomicLong();
+                try {
+                    writeRecords(recordReader, rowWriter, struct, recordsProcessed);
+                    session.adjustCounter(RECORDS_PROCESSED_COUNTER, recordsProcessed.get(), false);
+                } catch (final Exception e) {
+                    getLogger().error("Write Rows to Table [{}] failed {}", tableIdentifier, flowFile, e);
+                    abortWriter(rowWriter, tableIdentifier);
+                    relationship.set(FAILURE);
+                }
+            } catch (final Exception e) {
+                getLogger().error("Processing Records for Table [{}] failed {}", tableIdentifier, flowFile, e);
+                relationship.set(FAILURE);
```

Comment:
Shouldn't we abort the writer here too?

```suggestion
                getLogger().error("Processing Records for Table [{}] failed {}", tableIdentifier, flowFile, e);
                abortWriter(rowWriter, tableIdentifier);
                relationship.set(FAILURE);
```